File: Graph.java
import java.util.HashMap;
import java.util.LinkedList;

public class Graph {
    private HashMap<Integer, Node> nodeLookupMap = new HashMap<Integer, Node>();

    public static class Node {
        private int id;
        LinkedList<Node> adjacent = new LinkedList<Node>();

        private Node(int id) {
            this.id = id;
        }
    };
    
    private Node getNode(int id) {
        return nodeLookupMap.get(id);
    }

    public void addEdge(int source, int destination) {
        Node s = getNode(source);
        Node d = getNode(destination);
        s.adjacent.add(d);
    }
}

File: graph.py
from collections import defaultdict

class Graph:
    nodeMap = defaultdict(lambda: None)
    
    class Node:
        def __init__(self, id, adjacent=None):
            self.id = id
            self.adjacent = adjacent

    def getNode(self, id):
        return self.nodeMap[id]
    
    def addEdge(self, source, target):
        n = self.getNode(source)
        s = self.getNode(target)
        s.adjacent.add(s)
    
    def hasPathDFS(self, source, target):
        n = self.getNode(source)
        s = self.getNode(target)
        visited = set()
        return self.hasPathDFS(source, target, visited)

    def hasPathDFS(self, source, target, visited):
        if source in visited:
            return False

        
def main():
    graph = Graph()


main()